src/main/flight/pid.c

```diff
@@ -803,77 +802,92 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
-#if defined(USE_ABSOLUTE_CONTROL)
-            gyroTargetLow += acCorrection;
-            gyroTargetHigh += acCorrection;
-#endif
-            if (axis < FD_YAW) {
-                int itemOffset = (axis << 1);
-                DEBUG_SET(DEBUG_ITERM_RELAX, itemOffset++, gyroTargetHigh);
-                DEBUG_SET(DEBUG_ITERM_RELAX, itemOffset, gyroTargetLow);
+            const float gyroTarget = pt1FilterApply(&windupLpf[axis], currentPidSetpoint);
+            const float gyroHpf = fabsf(currentPidSetpoint - gyroTarget);
+            if (itermRelaxType == ITERM_RELAX_SETPOINT) {
+                itermErrorRate = (1 - MIN(1, fabsf(gyroHpf) / 60.0f)) * (currentPidSetpoint - gyroRate);
+            }
+            const float tolerance = gyroHpf * 1.0f;
+            if (axis == FD_ROLL) {
+                DEBUG_SET(DEBUG_ITERM_RELAX, 0, gyroTarget);
+                DEBUG_SET(DEBUG_ITERM_RELAX, 1, gyroTarget + tolerance);
+                DEBUG_SET(DEBUG_ITERM_RELAX, 2, gyroTarget - tolerance);
+                DEBUG_SET(DEBUG_ITERM_RELAX, 3, axisError[axis] * 10);
             }
-            const float gmax = MAX(gyroTargetHigh, gyroTargetLow);
-            const float gmin = MIN(gyroTargetHigh, gyroTargetLow);
-            if (gyroRate < gmin || gyroRate > gmax) {
-                itermErrorRate = (gyroRate > gmax ? gmax : gmin ) - gyroRate;
+            if (itermRelaxType == ITERM_RELAX_GYRO ) {
+                const float gmax = gyroTarget + tolerance;
+                const float gmin = gyroTarget - tolerance;
+                
+                if (gyroRate >= gmin && gyroRate <= gmax) {
+                     itermErrorRate = 0;
+                } else {
+                     itermErrorRate = (gyroRate > gmax ? gmax : gmin ) - gyroRate;
+                }
             }
+            
 #if defined(USE_ABSOLUTE_CONTROL)
-            else {
-                itermErrorRate = acCorrection;
+            const float gmaxac = gyroTarget + 2 * tolerance;
+            const float gminac = gyroTarget - 2 * tolerance;
+            if (gyroRate >= gminac && gyroRate <= gmaxac) {
+                float acErrorRate1 = gmaxac - gyroRate;
+                float acErrorRate2 = gminac - gyroRate;
+                if (acErrorRate1 * axisError[axis] < 0) {
+                    acErrorRate = acErrorRate1;
+                } else {
+                    acErrorRate = acErrorRate2;
+                }
+                if (fabsf(acErrorRate*dT) > fabsf(axisError[axis]) ) {
+                    acErrorRate = -axisError[axis]/dT;
+                }
+            } else {
+                acErrorRate = (gyroRate > gmaxac ? gmaxac : gminac ) - gyroRate;
             }
-            acErrorRate = itermErrorRate - acCorrection;
-#endif
+#endif // USE_ABSOLUTE_CONTROL             
         } else
 #endif // USE_ITERM_RELAX
         {
-            itermErrorRate = errorRate;
+              itermErrorRate = currentPidSetpoint - gyroRate;
```

Comment:
There seem to be more spaces than necessary?

Reply:
Fixed. Parts of this file seem to be having the wrong indentation, but I haven't cleaned this up for the sake of easy reviewing.